Infrastructure to enable container replacement

josecv opened the pull request:
## Description

This PR adds changes to the agent injector and traffic manager to enable the creation of intercepts that replace the application container. The CLI flag is not implemented here.

## Checklist

<!--
  Please review the requirements for each checkbox, and check them
  off (change "[ ]" to "[x]") as you verify that they are complete.
-->

 - [ ] I made sure to update `./CHANGELOG.yml`.
 - [ ] I made sure to add any docs changes required for my change (including release notes).
 - [ ] My change is adequately tested.
 - [ ] I updated `DEVELOPING.md` with any special dev tricks I had to use to work on this code efficiently.
 - [ ] I updated `TELEMETRY.md` if I added, changed, or removed a metric name.
 - [ ] Once my PR is ready to have integration tests ran, I posted the PR in #telepresence-dev in the datawire-oss slack so that the "ok to test" label can be applied.

josecv committed:
Delete app container if replacing it

Signed-off-by: Jose Cortes <josecortes@datawire.io>

josecv committed:
Consistent use of the context in the state struct

We should pass a context into every function going into state; it's the
request context.

Signed-off-by: Jose Cortes <josecortes@datawire.io>

petergardfjall commented:
@josecv Is this PR what I think it is, a `telepresence intercept --replace` command?
(That's almost exactly what I wished for [here](https://github.com/telepresenceio/telepresence/issues/1608#issuecomment-1091614626). I'm still stuck on v1 of telepresence since I have an application with work queue semantics).

Would love to see it completed!

josecv committed:
Run intercept finalizers in their own goroutine

AddInterceptFianlizer is called by stuff outside the state struct; hence
it makes no sense to allow a fianlizer to run holding the lock. If
needed, it can re-acquire it through the state object; the lock is only
protecting the state itself, i.e. the removal of the intercept from the
map.

This also requires some coordination to make sure that the calls into
the state object wait for finalizers to run. There's a new wait()
mechanism to enable this.

Signed-off-by: Jose Cortes <josecortes@datawire.io>

josecv committed:
Wire the replace parameter into the configmap

Signed-off-by: Jose Cortes <josecortes@datawire.io>

thallgren commented on rpc/manager/manager.proto:
```diff
@@ -75,50 +75,47 @@ message InterceptSpec {
   // Same as AgentInfo.Namespace of the Workload
   string namespace = 8;
 
-  // How to decide which subset of requests to that agent to intercept.
```
I'm not in favor or reordering based on the gRPC number for each attribute. That number is to ensure backward compatibility when things move around, and it's only used by gRPC internals.

I'm of course fine with reordering attributes, but when we do, then the order should either be alphabetical or based on logical grouping of attributes (both rules broken here, because `mechanism` is moved away from `mechanism_args`).

thallgren commented on pkg/agentconfig/sidecar.go:
```diff
@@ -153,6 +158,9 @@ type Sidecar struct {
 	InitResources *core.ResourceRequirements `json:"initResources,omitempty"`
 
+	// User-supplied configuration for the intercept.
+	UserConfig *UserConfig `json:"userConfig,omitempty"`
```
A `Sidecar` has many `Containers`, that has many `Intercepts`. The `--replace` flag affects an `Intercept`. So placing this in the `Sidecar` is a bit inconsistent. I realise that the flag must affect all `Intercepts` for a specific `Container`, because the affected `Container` is removed, but why must it affect all `Containers` managed by a `Sidecar`?

I would like to have a `Replace` attribute on the `Container` here, and no `UserInfo`. Once set, that attribute should be considered persistent.

Moving forward (not in this PR) I think it might be a good idea to also enable control over that attribute via an annotation in the workload's pod-template, because this information is really fairly static. It's not something that users are likely to set in different ways. Topic for discussion for sure.

josecv replied:
The problem is that the injector doesn't know whether or not the workload is currently intercepted. It tries to be idempotent with respect to the configmap. If The attribute is set persistently to true, then how does the agent know whether or not to do the replacement?

thallgren commented on cmd/traffic/cmd/manager/mutator/agent_injector.go:
```diff
@@ -230,6 +231,27 @@ func needInitContainer(config *agentconfig.Sidecar) bool {
 	return false
 }
 
+func deleteAppContainer(ctx context.Context, pod *core.Pod, config *agentconfig.Sidecar, patches patchOps) patchOps {
```
This should delete the containers that are marked for replace, not all containers that can be intercepted.

josecv committed:
Address PR comments

Signed-off-by: Jose Cortes <josecortes@datawire.io>

thallgren commented on cmd/traffic/cmd/manager/service.go:
```diff
@@ -443,7 +446,11 @@ func (s *service) PrepareIntercept(ctx context.Context, request *rpc.CreateInter
-	return s.state.PrepareIntercept(ctx, request)
+	replacePolicy := agentconfig.ReplacePolicyNever
+	if request.InterceptSpec.Replace {
+		replacePolicy = agentconfig.ReplacePolicyInactive
```
Shouldn't this be ReplacePolicyActive?

josecv replied:
Nope; it's just the prepare call, the intercept isn't live yet. The CreateIntercept call will then switch it to Active. This way Prepare is just used for the setup of the configmap, i.e. to ensure that it's impossible to later say --replace=false for instance

josecv replied:
If this were Active, I think it'd violate the API for PrepareIntercpept. After all, then it'd be possible to call Prepare without a subsequent Create and effectively bring down the service (since there's no listener for the agent to relay the request to)

thallgren commented on pkg/agentconfig/sidecar.go:
```diff
@@ -49,6 +49,17 @@ const (
+	ReplacePolicyNever ReplacePolicy = iota
+	// --replace is true, the intercept is active.
+	ReplacePolicyActive
+	// -replace is true, the intercept is inactive.
```
```suggestion
	// --replace is true, the intercept is inactive.
```

josecv committed:
Comment fix

Co-authored-by: Thomas Hallgren <thomas@datawire.io>
Signed-off-by: Jose Cortes <josecortes@datawire.io>

petergardfjall commented:
Any ETA for a `--replace` CLI flag? :star_struck:
